faunadb-java-dsl/src/main/java/com/faunadb/client/types/Constructors.java

```diff
@@ -0,0 +1,211 @@
+package com.faunadb.client.types;
+
+import com.faunadb.client.errors.FaunaException;
+import com.faunadb.client.types.Properties.Property;
+import com.faunadb.client.types.Types.SimpleType;
+import com.faunadb.client.types.Value.ObjectV;
+
+import java.lang.annotation.Annotation;
+import java.lang.reflect.Constructor;
+import java.lang.reflect.Method;
+import java.lang.reflect.Modifier;
+import java.lang.reflect.Type;
+
+import static com.faunadb.client.types.Decoder.decodeImpl;
+import static java.lang.String.format;
+
+final class Constructors {
+    static Codec<Object> createCodec(Class<?> clazz) {
+        Codec<Object> codec = getStaticMethodCreatorCodec(clazz);
+
+        if (codec == null)
+            codec = getAnnotatedConstructorCodec(clazz);
+
+        if (codec == null)
+            codec = getDefaultConstructorCodec(clazz);
+
+        if (codec != null)
+            return codec;
+
+        throw new FaunaException(format("No suitable constructor/creator found for type %s", clazz));
+    }
+
+    private static Codec<Object> getStaticMethodCreatorCodec(Class<?> clazz) {
+        for (Method method : clazz.getDeclaredMethods()) {
+            if (!method.isAnnotationPresent(FaunaConstructor.class))
+                continue;
+
+            if ((method.getModifiers() & (Modifier.STATIC | Modifier.PUBLIC)) == 0)
+                continue;
+
+            if (method.getParameterTypes().length == 0)
+                return new NoArgsStaticMethodCodec(method);
+
+            return new StaticMethodCodec(method);
+        }
+
+        return null;
+    }
+
+    private static Codec<Object> getAnnotatedConstructorCodec(Class<?> clazz) {
+        for (Constructor<?> constructor : clazz.getDeclaredConstructors()) {
+            if (constructor.isAnnotationPresent(FaunaConstructor.class)) {
+                if (constructor.getParameterTypes().length == 0)
+                    return new DefaultConstructorCodec(constructor);
+
+                return new ConstructorCodec(constructor);
+            }
+        }
+
+        return null;
+    }
+
+    private static Codec<Object> getDefaultConstructorCodec(Class<?> clazz) {
+        try {
+            return new DefaultConstructorCodec(clazz.getDeclaredConstructor());
+        } catch (NoSuchMethodException ex) {
+            return null;
+        }
+    }
+
+    static private abstract class AbstractCreator implements Codec<Object> {
```

Comment:
I think we should replace all references to `Creator` with `Constructor` just for consistency.